### PR TITLE
feat(calling): Adding cluster redundancy for discovery failure

### PR DESCRIPTION
**Pull request by Kesari3008**
# COMPLETES #https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-470778

## This pull request addresses
Mobius cluster redundancy for discovery failure scenarios

## by making the following changes
Stored _hostCatalog within Webex object in the service catalog.
Fetching Mobius cluster entries from _hostCatalog within Webex object and traversing through each host in case of discovery failures.
If the request to /myIp fails or the region request fails on the returned IP then we will check the next entry from the Mobius host array and repeat the same process. The Mobius cluster for which fetching the IP address followed by region discovery is successful then for the same cluster primary and backup Mobius URI are fetched and used for device registration.

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [x] Internal code refactor

## The following scenarios where tested
Testcase 1: Pass the region and country in the discovery config and proceed with discovery followed by device registration. 
[Discovery_with_config_console_logs.txt](https://github.com/webex/webex-js-sdk/files/13423010/Discovery_with_config_console_logs.txt)
[Discovery_With_Config_network_logs.txt](https://github.com/webex/webex-js-sdk/files/13423011/Discovery_With_Config_network_logs.txt)

Testcase 2: Replicate the failed response for myIP request and verify that it traverses through each entry present in the mobius cluster array and attempts the discovery.
[Discovery_failure_console_logs.log](https://github.com/webex/webex-js-sdk/files/13423229/Discovery_failure_console_logs.log)
[Discovery_failure_network_logs.txt](https://github.com/webex/webex-js-sdk/files/13423238/Discovery_failure_network_logs.txt)

Testcase3: Discovery request receives success response
[Discovery_Success_logs_console_next.log](https://github.com/webex/webex-js-sdk/files/13478718/Discovery_Success_logs_console_next.log)
[Discovery_success_har_next.txt](https://github.com/webex/webex-js-sdk/files/13478716/Discovery_success_har_next.txt)

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [x] I discussed changes with code owners prior to submitting this pull request

- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [ ] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/@webex/webex-core/src/lib/services/services.js [hunk 1]
```diff
@@ -959,6 +971,7 @@ const Services = WebexPlugin.extend({
           // Validate if the token is authorized.
           if (credentials.canAuthorize) {
             // Attempt to collect the postauth catalog.
+
```
Comment (sreenara) [inline packages/@webex/webex-core/src/lib/services/services.js:974]: this change is not needed

File: packages/@webex/webex-core/src/lib/services/services.js [hunk 2]
```diff
@@ -166,6 +168,15 @@ const Services = WebexPlugin.extend({
     this._serviceUrls = {...this._serviceUrls, ...serviceUrls};
   },
 
+  /**
+   * saves the host catalog
+   * @param {Object} hostCatalog
+   * @returns {void}
+   */
+  _updateHostCatalog(hostCatalog) {
```
Comment (sreenara) [inline packages/@webex/webex-core/src/lib/services/services.js:176]: do we not have any changes to tests?

File: packages/calling/src/CallingClient/CallingClient.test.ts [hunk 3]
```diff
@@ -254,31 +257,156 @@ describe('CallingClient Tests', () => {
...
+
       expect(warnSpy).toBeCalledWith(
-        'Error in finding Mobius Servers. Will use the default URL.',
+        `Couldn't resolve the region and country code. Defaulting to the catalog entries to discover mobius servers`,
         ''
       );
     });
 
+    it('verifying entries of eu mobius clusters', async () => {
+      const mockEUServiceHosts = [
```
Comment (sreenara) [inline packages/calling/src/CallingClient/CallingClient.test.ts:300]: can we move these to a fixtures file so we don't have so many extra lines in the test file?

File: packages/calling/src/CallingClient/CallingClient.test.ts [hunk 4]
```diff
@@ -254,31 +257,156 @@ describe('CallingClient Tests', () => {
...
+          id: 'urn:TEAM:us-east-2_a:mobius',
+        },
+        {
+          host: 'mobius-ap-southeast-2.int.infra.webex.com',
+          ttl: -1,
+          priority: 20,
+          id: 'urn:TEAM:us-east-2_a:mobius',
+        },
+      ];
+      delete webex.internal.services._hostCatalog['mobius-us-east-1.prod.infra.webex.com'];
```
Comment (sreenara) [inline packages/calling/src/CallingClient/CallingClient.test.ts:399]: is there an easier way to cover all the entries than deleting the previous entries? Perhaps we could have 4 constants where EU central would have highest priority, then US EAST, then CA CENTRAL etc and then just do it.each across those 4 constants with the same test code?

**Review by sreenara — CHANGES_REQUESTED** [review]
1. Please see the comments.  
2. Please change the title of the PR.
3. Please add the test case of a successful region discovery along with registration like the web-calling-sdk repo

File: packages/@webex/webex-core/src/lib/services/services.js (hunk 2 above)
Comment (Kesari3008) [reply packages/@webex/webex-core/src/lib/services/services.js:176]: we do...I missed adding it here in the next branch. Added now

File: packages/@webex/webex-core/src/lib/services/services.js (hunk 1 above)
Comment (Kesari3008) [reply packages/@webex/webex-core/src/lib/services/services.js:974]: removed

File: packages/calling/src/CallingClient/CallingClient.test.ts (hunk 3 above)
Comment (Kesari3008) [reply packages/calling/src/CallingClient/CallingClient.test.ts:300]: done

File: packages/calling/src/CallingClient/CallingClient.test.ts (hunk 4 above)
Comment (Kesari3008) [reply packages/calling/src/CallingClient/CallingClient.test.ts]: done

File: packages/calling/src/common/testUtil.ts [hunk 5]
```diff
@@ -7,6 +7,12 @@
 */
 import {MediaConnectionConfig} from '@webex/internal-media-core';
 import platform from 'platform';
+import {
+  mockEUIntServiceHosts,
+  mockEUServiceHosts,
+  mockIntServiceHosts,
+  mockUSServiceHosts,
+} from 'CallingClient/callingClientFixtures';
```
Comment (sreenara) [inline packages/calling/src/common/testUtil.ts:15]: please fix the path. This will cause issues.

**Review by sreenara — CHANGES_REQUESTED** [review]
Please change the title to match the usual template and add the logs with successful discovery as mentioned in the previous comments.

File: packages/calling/src/CallingClient/CallingClient.test.ts [hunk 6]
```diff
@@ -342,13 +368,47 @@ describe('CallingClient Tests', () => {
...
   });
 
+  describe('Testing each cluster present withing host catalog', () => {
+    const mobiusCluster = [
+      'mobius-eu-central-1.prod.infra.webex.com',
+      'mobius-us-east-1.int.infra.webex.com',
+      'mobius-eu-central-1.int.infra.webex.com',
+    ];
+
+    const checkCluster = async (
```
Comment (sreenara) [inline packages/calling/src/CallingClient/CallingClient.test.ts:385]: Nice work on the test

**Review by sreenara — APPROVED** [review]

**Pull request merged**